velox/docs/programming-guide/chapter02.rst

```diff
@@ -0,0 +1,356 @@
+    > 0: RED
+      1: GREEN
+      2: BLUE
+      3: RED
+      4: GREEN
+
+The vector now has non-empty strings, but still no string buffers. That’s
+because we wrote only short strings (<= 12 bytes each). These strings are fully
+stored within the StringViews inside the values buffer and do not require
+separate string buffers.
+
+Let’s now write some long strings.
+
+.. code-block:: c++
+
+    for (auto i = 0; i < 100; ++i) {
+     if (i % 3 == 0) {
+       vector->set(i, "In my hometown where I used to stay");
+     } else if (i % 3 == 1) {
+       vector->set(i, "The name of the place is Augusta, GA");
+     } else if (i % 3 == 2) {
+       vector->set(i, "Down there we have a good time");
+     }
+    }
+
+    LOG(INFO) << "Number of string buffers: " << vector->stringBuffers().size();
+    LOG(INFO) << vector->toString(0, 10);
+
+    > Number of string buffers: 1
+    > 0: In my hometown where I used to stay
+      1: The name of the place is Augusta, GA
+      2: Down there we have a good time
+      3: In my hometown where I used to stay
+      4: The name of the place is Augusta, GA
+
+    const auto& stringBuffer = vector->stringBuffers()[0];
+    LOG(INFO) << stringBuffer->size() << ", " << stringBuffer->capacity();
+
+    > 3368, 49056
+
+    LOG(INFO) << std::string_view(stringBuffer->as<char>(), 100);
+
+    > In my hometown where I used to stayThe name of the place is Augusta, GADown there we have a good tim
+
+Now, we see that there is one string buffer of size 3368 bytes and total
+capacity of 49056 bytes. We can also see that this buffer contains the strings
+we wrote one after another.
+
+If we change some strings, the new strings will be written to the end of the
+buffer and if we exceed the buffer's capacity, a new buffer will be allocated.
+
+.. code-block:: c++
+
+    for (auto n = 0; n < 25; ++n) {
+        for (auto i = 0; i < 100; i += 2) {
+            vector->set(i, "We all get together in time, for rhythm then we do");
+        }
+
+        LOG(INFO) << n << ": Number of string buffers: "
+                  << vector->stringBuffers().size();
+
+        for (const auto& buffer : vector->stringBuffers()) {
+          LOG(INFO) << buffer->size() << ", " << buffer->capacity();
+        }
+    }
+
+    > 0: Number of string buffers: 1
+    > 5868, 49056
+    ...
+    > 17: Number of string buffers: 1
+    > 48368, 49056
+
+    > 18: Number of string buffers: 2
+    > 49018, 49056
+    > 1850, 49056
+
+FlatVector<StringView>::set(index, value) method copies the string to the end of
+a string buffer and updates StringView in the values buffer to point to that
+copy. The old string is not removed and still occupies space in the buffer. We
+need to be careful not to end up with many buffers full of unreferenced
+strings. Also, FlatVector<StringView>::set doesn't check if the "new" string
+already exists. There is no de-duplication logic.
+
+We can also provide our own string buffers and make StringViews that refer to
```

Comment:
Can we have strings that share parts of the buffer? something like
```
0: In my hometown where I used to stay
1: hometown where I used to stay
```
If yes, we should discourage any inplace modifications to the string buffer and stick to append only strategy like we do in the set() API

Reply:
I believe it is possible to use overlapping strings. Strings in the buffers are considered immutable though... This is something that would be nice to clarify.